src/bitmap/store.rs

```diff
@@ -136,28 +151,36 @@ impl Store {
         }
     }
 
-    pub fn remove_range(&mut self, start: u32, end: u32) -> u64 {
-        debug_assert!(start < end, "caller must ensure start < end");
+    pub fn remove_range(&mut self, range: RangeInclusive<u16>) -> u64 {
+        if range.is_empty() {
+            return 0;
+        }
+
+        let start = *range.start();
+        let end = *range.end();
+
         match *self {
             Array(ref mut vec) => {
-                let a = vec.binary_search(&(start as u16)).unwrap_or_else(|e| e);
-                let b = if end > u32::from(u16::max_value()) {
-                    vec.len()
-                } else {
-                    vec.binary_search(&(end as u16)).unwrap_or_else(|e| e)
-                };
-                vec.drain(a..b);
-                (b - a) as u64
+                // Figure out the starting/ending position in the vec.
+                let pos_start = vec.binary_search(&start).unwrap_or_else(|x| x);
+                let pos_end = vec
+                    .binary_search_by(|p| {
+                        // binary search the right most position when equals
+                        match p.cmp(&end) {
+                            Greater => Greater,
+                            _ => Less,
+                        }
+                    })
+                    .unwrap_or_else(|x| x);
+                vec.drain(pos_start..pos_end);
+                (pos_end - pos_start) as u64
             }
             Bitmap(ref mut bits) => {
-                let start_key = key(start as u16) as usize;
-                let start_bit = bit(start as u16) as u32;
-                // end_key is inclusive
-                let end_key = key((end - 1) as u16) as usize;
-                let end_bit = bit(end as u16) as u32;
+                let (start_key, start_bit) = (key(start), bit(start));
+                let (end_key, end_bit) = (key(end), bit(end));
 
                 if start_key == end_key {
-                    let mask = (!0u64 << start_bit) & (!0u64).wrapping_shr(64 - end_bit);
+                    let mask = (!0u64 << start_bit) & (!0u64 >> (63 - end_bit));
```

Comment:
I'm not sure if `!0u64` is necessary, since `!0u64` = `u64::MAX`, and the later one is much more straightforward.